lib/bundler/definition.rb

```diff
@@ -800,10 +800,14 @@ def converge_locked_specs
           # commonly happens if the version changed in the gemspec
           next unless other
 
-          deps2 = other.dependencies.select {|d| d.type != :development }
+          new_deps = other.dependencies.select {|d| d.type != :development }
           runtime_dependencies = s.dependencies.select {|d| d.type != :development }
-          # If the dependencies of the path source have changed, unlock it
-          next unless runtime_dependencies.sort == deps2.sort
+          # If the dependencies of the path source have changed and locked spec can't satisfy new dependencies, unlock it
+          if runtime_dependencies.sort != new_deps.sort && new_deps.any? {|d| !satisfies_locked_spec?(d) }
+            next
+          else
+            s.dependencies.replace(other.dependencies)
+          end
```

Comment:
I think the patch would be easier to understand if it read like this:

```diff
 
           deps2 = other.dependencies.select {|d| d.type != :development }
           runtime_dependencies = s.dependencies.select {|d| d.type != :development }
-          # If the dependencies of the path source have changed, unlock it
-          next unless runtime_dependencies.sort == deps2.sort
+          # If the dependencies of the path source have changed and locked spec can't satisfy new dependencies, unlock it
+          next unless runtime_dependencies.sort == deps2.sort || deps2.all? {|d| satisfies_locked_spec?(d) }
+          s.dependencies.replace(other.dependencies)
         end
 
         converged << s
```

which I think it's equivalent?